### PR TITLE
Add a newAttachHeadersServerInterceptor() util

### PR DESCRIPTION
**Pull request by jdcormie**

### CODE REVIEW

File: stub/src/main/java/io/grpc/stub/MetadataUtils.java [hunk 1]
```diff
@@ -143,4 +147,41 @@ public void onClose(Status status, Metadata trailers) {
       }
     }
   }
+
+  /**
+   * Returns a ServerInterceptor that attaches a given set of headers to every response.
+   *
+   * @param extraHeaders the headers to be added to each response. Caller gives up ownership.
+   */
+  public static ServerInterceptor newAttachHeadersServerInterceptor(Metadata extraHeaders) {
```
Comment (ejona86) [inline stub/src/main/java/io/grpc/stub/MetadataUtils.java:156]: `@ExperimentalApi`

**Review by ejona86 — APPROVED** [review]

File: stub/src/main/java/io/grpc/stub/MetadataUtils.java (hunk 1 above)
Comment (jdcormie) [reply stub/src/main/java/io/grpc/stub/MetadataUtils.java:158]: done

File: stub/src/main/java/io/grpc/stub/MetadataUtils.java [hunk 2]
```diff
@@ -143,4 +148,42 @@ public void onClose(Status status, Metadata trailers) {
...
+    }
+
+    final class HeaderAttachingServerCall<ReqT, RespT>
+        extends SimpleForwardingServerCall<ReqT, RespT> {
+      HeaderAttachingServerCall(ServerCall<ReqT, RespT> delegate) {
+        super(delegate);
+      }
+
+      @Override
+      public void sendHeaders(Metadata headers) {
```
Comment (ejona86) [inline stub/src/main/java/io/grpc/stub/MetadataUtils.java:183]: FYI: If the response is trailers-only (generally an error), this won't attach the user's headers.

Comment (jdcormie) [reply stub/src/main/java/io/grpc/stub/MetadataUtils.java:185]: Fixed. Thanks for noticing.

File: stub/src/main/java/io/grpc/stub/MetadataUtils.java [hunk 3]
```diff
@@ -143,4 +148,56 @@ public void onClose(Status status, Metadata trailers) {
...
+        super.sendHeaders(headers);
+      }
+
+      @Override
+      public void close(Status status, Metadata trailers) {
+        if (!headersSent) {
+          // It isn't too late to call sendHeaders(): !headersSent implies that it hasn't been
+          // called yet (obviously). But it also implies that no messages have been sent, because
+          // sendMessage() *requires* a preceding call to sendHeaders().
+          sendHeaders(new Metadata());
```
Comment (ejona86) [inline stub/src/main/java/io/grpc/stub/MetadataUtils.java:197]: Oh, you're forcing headers. I don't think we want to do that generally. Maybe that is what you need, but I think it then stops being a general util.

Trailers-only is important to retries, as it delivers the status code without committing.

Comment (jdcormie) [reply stub/src/main/java/io/grpc/stub/MetadataUtils.java]: I didn't know that - fixed and thanks!

**Pull request merged**